xds/internal/resolver/serviceconfig_action.go

```diff
@@ -0,0 +1,158 @@
+//
+// Note that this only handles actions, but not matches. The routes in the
+// generated balancer config should maintain the same order as in RDS resp.
+// A separate list of routes would work.
+//
+
+// newActionsFromRoutes gets actions from the routes, and turns them into a map
+// keyed by the hash of the clusters.
+//
+// In the returned map, all actions don't have assignedName. The assignedName
+// will be filled in after comparing the new actions with the existing actions,
+// so when a new and old action only diff in weights, the new action can reuse
+// the old action's name.
+//
+// from
+//   {B:60, A:40}, {A:30, B:70}, {B:90, C:10}
+//
+// to
+//   A40_B60_: {{A:40, B:60}, "A_B_", ""}
+//   A30_B70_: {{A:30, B:70}, "A_B_", ""}
+//   B90_C10_: {{B:90, C:10}, "B_C_", ""}
+func newActionsFromRoutes(routes []*xdsclient.Route) map[string]action {
+	newActions := make(map[string]action)
+
+	for _, route := range routes {
+		var clusterNames []string
+		clusters := route.Action
+		for n := range clusters {
+			clusterNames = append(clusterNames, n)
+		}
+
+		// Sort names to be consistent.
+		sort.Strings(clusterNames)
+		clustersOnly := ""
+		clustersWithWeight := ""
+		for _, c := range clusterNames {
+			// Generates A_B_
+			clustersOnly = clustersOnly + c + "_"
+			// Generates A40_B60_
+			clustersWithWeight = clustersWithWeight + c + strconv.FormatUint(uint64(clusters[c]), 10) + "_"
+		}
+
+		if _, ok := newActions[clustersWithWeight]; !ok {
+			newActions[clustersWithWeight] = action{
+				clustersWithWeights: clusters,
+				clusterNames:        clustersOnly,
+			}
+		}
+	}
+	return newActions
+}
+
+// updateActions takes a new map of actions, and updates the existing action map in the resolver.
+//
+// In the old map, all actions have assignedName set.
+// In the new map, all actions have no assignedName.
+//
+// After the update, the action map is updated to have all actions from the new
+// map, with assignedName:
+// - if the new action exists in old, get the old name
+// - if the new action doesn't exist in old
+//   - if there is an old action that will be removed, and has the same set of
+//   clusters, reuse the old action's name
+//   - otherwise, generate a new name
+func (r *xdsResolver) updateActions(newActions map[string]action) {
+	// Delete actions from oldActions if they are not in newActions.
+	// Keep the removed actions in a map, with key as clusterNames without
+	// weights, so their assigned names can be reused.
+	if r.actions == nil {
+		r.actions = make(map[string]action)
+	}
+	oldActions := r.actions
+	actionsRemoved := make(map[string]action)
+	for clusterNamesWithWeight, act := range oldActions {
+		if _, ok := newActions[clusterNamesWithWeight]; !ok {
+			actionsRemoved[act.clusterNames] = act
+			delete(oldActions, clusterNamesWithWeight)
+		}
+	}
+
+	// Find actions in newActions but not in oldActions. Add them, and try to
+	// reuse assigned names from actionsRemoved.
+	if r.nextIndex == nil {
```

Comment:
Does this `nextIndex` thing need to be a monotonically increasing number? How do we protect against integer overall (even though that is an extremely rare event)?

Why not use a random number for the index part? We wouldn't have to maintain a per `clustersOnlyName` map for such an index. The random number would be over all `clustersOnlyName`s. We would still need to watch for hash collision, but that would simply involve generating another random number.

Reply:
I'm not too worried about overflow, and I made it a `uint64` now.

Random number would work, but it makes the tests harder.

Reply:
We could override the random generator in tests with something that returns monotonically increasing values for the call to `Int63n`, can't we?
I just feel the logic to maintain this `nextIndex` on a per clusterHashName seems excessive, given that the actual number returned by the `nextIndex` does not have much significance except that it needs to be unique.
I don't see this come up in the design doc. Was this something that was discussed?

Reply:
Updated to use random number.
I do need to change some tests so the output is expected, but it isn't too bad.